README.md
# ManyEx
System to create examinations in LaTeX with permutations from a database of exercises and questions

manyex: a program to generate exams using LaTeX and Tth
=======================================================

Linux/Unix
----------

To install simply run "make" and (as root) "make install". Adjustments in the
makefile may be necessary (e.g. the INSTALLDIR variable may need to be
changed).


Examples
--------

To run a simple example, just enter the "example_simple" directory and
enter the following command line:
manyex -n -f desc-test

This will create the files "exam.tex", with the exams, "answer.tex"
formatted answers, "mn-correct" with answer key and "mn-logfile" with
a log of the operations performed.

A more complex example can be run from the example_normal directory. In
this case almost all capabilities for normal LaTeX exam creation can
be tested. See the test file description in "desc-test" and the master
LaTeX file in that directory.

An example of web form creation can be found in example_web. Please check
the files desc-test and master.sty in that subdirectory. To create
the LaTeX web form exams give the following command:
manyex -w -f desc-test

To create the html forms you need to download tth and filter the latex
output through this program. If you want to actually use the forms you
need access to a web server and knowledge on scripting to process the forms.

Shortcomings and limitations
----------------------------

Error checking is quite limited for the exercise syntax. The syntax for
exercises has to be rewritten completely to improve it, right now it works
but syntax errors in the exercise definition produce unpredictable results.
For instance commands in the execise files cannot have white space after
the command and have to start at the beginning of the line. An error
consisting of white space after a command in the exercise file is very
difficult to debug, since the program reports an error at the end of the file.

Quite large strings are defined in vars.h. If compilation is succesful but
the program fails to run most likely there is not enough memory to hold
the strings. The size can be reduced in vars.h, making sure that no exercise
exceeds these limits. String operation has to be rewritten, it works but
the program is very inefficient.


ChangeLog
Version 0.9
2010-06-15
	* Fixed bug 5: regression made statements not print in answers

Version 0.8
2010-06-15
	* Fixed bug 4: statements repeated in other exams

Version 0.7
2010-06-15
	* Introduced option to make fields required in web exams
	* Fixed a couple of bugs

Version 0.6
2008-08-22
	* Fixed a bug in create.c that was creating a stack overflow in 
	Linux, but surprisingly not in Cygwin
	* Removed a couple of routines fromm randomlib.c which were
	causing some warnings but were not being used

Version 0.5
2006-10-29
	* Changed code for random selection of exercises, so that the
	exams use all of the exercises available in the database and do
	not repeat the same exercise unless needed

Version 0.4
2006-10-22
	* Introduced uncheck boxes in webform multiplechoice exams

Version 0.3
2006-10-11
	* Fixed some minor bugs introduced in the previos release

Version 0.2
2006-10-11
	* Fixed bug that make title of exercise continue appearing in new
	exercises if no title was included after title was defined

VERSION 0.1
2006-09-04
	* Added some comments on shortcomings and limitations in README

VERSION 0.0.9
2006-09-04	
	* Fixed bug in exer.y, the value of the variable rearrange was
	incorrectly set for some cases

VERSION 0.0.8
2006-07-29	
	* Fixed bug in exer.y, web integer was not initialized after a block

VERSION 0.0.7
2006-07-28
	* corrected some minor things in manyex.texi

VERSION 0.0.6
2006-07-27
	* added a complete example of html form exam
	* completeed the documentation on a complete example
	* fixed in create.c: processing of latex input in options

VERSION 0.0.5
2006-07-25
	* added a complete example of a normal LaTeX exam
	* started working on the documentatino of a complete example
	* fixed line numbering for errors in exercise files
	* modified a couple of things in Makefile 
	* fixed master.sty in example_simple, \mnanswerpaperheader was not 
	defined
	
VERSION 0.0.4
2006-07-25
	* added example_base and example_normal, complete normal example
	* modified README to give directions for complete normal example

VERSION 0.0.3
2006-07-24
	* added a ";" at the end of the rules block in exer.y and desc.y,
	which was giving an error in some systems
	* changed MAX_STRING_SIZE to 200 instead of 1000


util.c
/*
  manyex: a program to produce exams from a database of exercises

  History:

  The inspiration and some code comes for the program "much":
  Copyright (C) 2004  Mihail N. Kolountzakis

  This program:
  Copyright (C) 2006-2010  Walter Garcia-Fontes

  Read the file COPYRIGHT for the complete copyright.

  Various utility subroutines.
*/

#include	"vars.h"
#define         BSIZE 512

// Substring function used in yacc for various things
substr(char dest[], char src[], int offset, int len)
{
  int i;
  for(i = 0; i < len && src[offset + i] != '\0'; i++)
    dest[i] = src[i + offset];
  dest[i] = '\0';
}

// generate a random subset of k elements of the set {0, ..., N-1} and
// return the elements in the vector elem
void    random_set(int N, int k, int *elem)
{
  int     n=0, i, r;
  char    *s;
  double  x;

  s = (char*)malloc(N);
  for(i=0; i<N; i++) s[i]=0;

  while(n<k) {
    x = RandomUniform();
    // x = drand48();
    x *= N;
    r = (int)x;
    if(s[r]) continue;
    s[r] = 1;
    elem[n++] = r;
  }
  free(s);
}

/*
  Simple function to copy the auxiliary files from the database to the 
  current directory. TODO: improve error handling in this file
*/
copyfile(from, to)
     char *from, *to;
{
struct stat stbuf1, stbuf2;
char iobuf[BSIZE];
  int fold, fnew, n;
  register char *p1, *p2, *bp;
  int mode;
  if ((fold = open(from, 0)) < 0) {
    fprintf(stderr, "cp: cannot open %s\n", from);
    return(1);
  }
  fstat(fold, &stbuf1);
  mode = stbuf1.st_mode;
  /* is target a directory? */
  if (stat(to, &stbuf2) >=0 &&
      (stbuf2.st_mode&S_IFMT) == S_IFDIR) {
    p1 = from;
    p2 = to;
    bp = iobuf;
    while(*bp++ = *p2++)
      ;
    bp[-1] = '/';
    p2 = bp;
    while(*bp = *p1++)
      if (*bp++ == '/')
	bp = p2;
    to = iobuf;
  }
  if (stat(to, &stbuf2) >= 0) {
    if (stbuf1.st_dev == stbuf2.st_dev &&
	stbuf1.st_ino == stbuf2.st_ino) {
      fprintf(stderr, "cp: cannot copy file to itself.\n");
      return(1);
    }
  }
  if ((fnew = creat(to, mode)) < 0) {
    fprintf(stderr, "cp: cannot create %s\n", to);
    close(fold);
    return(1);
  }
  while(n = read(fold,  iobuf,  BSIZE)) {
    if (n < 0) {
      fprintf(stderr, "cp: read error\n");
      close(fold);
      close(fnew);
      return(1);
    } else
      if (write(fnew, iobuf, n) != n) {
	fprintf(stderr, "cp: write error.\n");
	close(fold);
	close(fnew);
	return(1);
      }
  }
  close(fold);
  close(fnew);
  return(0);
}




// Simple function to convert integer to string
char *itoa(int a,char *buffer, int buff_size) 
{ 
  char *temp; 
  int b; 
  temp = buffer; 

  b = a; 
  if (buffer == NULL) 
    return NULL; 

  while (a > 0) 
    { 
      if (buff_size < 1) 
	return NULL; 

      b = a%10; 
      *temp = b + '0' ; 
      temp++; 
      a = a/10; 
      buff_size--; 
    } 
  *temp = '\0'; 
  /*Note : You have to Add a String reverse Function on this buffer here*/ 
  return buffer; 
} 



randomlib.c
#define FALSE 0
#define TRUE 1

/*
   This Random Number Generator is based on the algorithm in a FORTRAN
   version published by George Marsaglia and Arif Zaman, Florida State
   University; ref.: see original comments below.
   At the fhw (Fachhochschule Wiesbaden, W.Germany), Dept. of Computer
   Science, we have written sources in further languages (C, Modula-2
   Turbo-Pascal(3.0, 5.0), Basic and Ada) to get exactly the same test
   results compared with the original FORTRAN version.
   April 1989
   Karl-L. Noell <NOELL@DWIFH1.BITNET>
      and  Helmut  Weber <WEBER@DWIFH1.BITNET>

   This random number generator originally appeared in "Toward a Universal
   Random Number Generator" by George Marsaglia and Arif Zaman.
   Florida State University Report: FSU-SCRI-87-50 (1987)
   It was later modified by F. James and published in "A Review of Pseudo-
   random Number Generators"
   THIS IS THE BEST KNOWN RANDOM NUMBER GENERATOR AVAILABLE.
   (However, a newly discovered technique can yield
   a period of 10^600. But that is still in the development stage.)
   It passes ALL of the tests for random number generators and has a period
   of 2^144, is completely portable (gives bit identical results on all
   machines with at least 24-bit mantissas in the floating point
   representation).
   The algorithm is a combination of a Fibonacci sequence (with lags of 97
   and 33, and operation "subtraction plus one, modulo one") and an
   "arithmetic sequence" (using subtraction).

   Use IJ = 1802 & KL = 9373 to test the random number generator. The
   subroutine RANMAR should be used to generate 20000 random numbers.
   Then display the next six random numbers generated multiplied by 4096*4096
   If the random number generator is working properly, the random numbers
   should be:
           6533892.0  14220222.0  7275067.0
           6172232.0  8354498.0   10633180.0
*/

/* Globals */
double u[97],c,cd,cm;
int i97,j97;
int test = FALSE;

/*
   This is the initialization routine for the random number generator.
   NOTE: The seed variables can have values between:    0 <= IJ <= 31328
                                                        0 <= KL <= 30081
   The random number sequences created by these two seeds are of sufficient
   length to complete an entire calculation with. For example, if sveral
   different groups are working on different parts of the same calculation,
   each group could be assigned its own IJ seed. This would leave each group
   with 30000 choices for the second seed. That is to say, this random
   number generator can create 900 million different subsequences -- with
   each subsequence having a length of approximately 10^30.
*/
void RandomInitialise(int ij,int kl)
{
   double s,t;
   int ii,i,j,k,l,jj,m;

   /*
      Handle the seed range errors
         First random number seed must be between 0 and 31328
         Second seed must have a value between 0 and 30081
   */
   if (ij < 0 || ij > 31328 || kl < 0 || kl > 30081) {
		ij = 1802;
		kl = 9373;
   }

   i = (ij / 177) % 177 + 2;
   j = (ij % 177)       + 2;
   k = (kl / 169) % 178 + 1;
   l = (kl % 169);

   for (ii=0; ii<97; ii++) {
      s = 0.0;
      t = 0.5;
      for (jj=0; jj<24; jj++) {
         m = (((i * j) % 179) * k) % 179;
         i = j;
         j = k;
         k = m;
         l = (53 * l + 1) % 169;
         if (((l * m % 64)) >= 32)
            s += t;
         t *= 0.5;
      }
      u[ii] = s;
   }

   c    = 362436.0 / 16777216.0;
   cd   = 7654321.0 / 16777216.0;
   cm   = 16777213.0 / 16777216.0;
   i97  = 97;
   j97  = 33;
   test = TRUE;
}

/* 
   This is the random number generator proposed by George Marsaglia in
   Florida State University Report: FSU-SCRI-87-50
*/
double RandomUniform(void)
{
   double uni;

   /* Make sure the initialisation routine has been called */
   if (!test) 
   	RandomInitialise(1802,9373);

   uni = u[i97-1] - u[j97-1];
   if (uni <= 0.0)
      uni++;
   u[i97-1] = uni;
   i97--;
   if (i97 == 0)
      i97 = 97;
   j97--;
   if (j97 == 0)
      j97 = 97;
   c -= cd;
   if (c < 0.0)
      c += cm;
   uni -= c;
   if (uni < 0.0)
      uni++;

   return(uni);
}




vars.h
/*
  manyex: a program to produce exams from a database of exercises

  History:

  The inspiration and some code comes for the program "much":
  Copyright (C) 2004  Mihail N. Kolountzakis

  This program:
  Copyright (C) 2006-2010  Walter Garcia-Fontes

  Read the file COPYRIGHT for the complete copyright.

  Include, macros and general definitions.
*/


#include        <stdio.h>
#include        <ctype.h>
#include        <string.h>
#include        <locale.h>
#include        <fnmatch.h>
#include        <dirent.h>
#include        <stdlib.h>
#include        <math.h>            
#include        <stdarg.h>
#include        <getopt.h>
#include        <errno.h>
#include        <sys/types.h> // needed for copyfile subroutine
#include        <sys/stat.h> // needed for copyfile subroutine

#include	"randomlib.h"


#ifdef  MAIN_PROGRAM
#define EXTERN /**/
#else
#define EXTERN extern
#endif

#define	MAX_STR_LENGTH	        200  // max string length of lines
#define MAX_STR_LABEL_TF        50 // max string for truefalse labels
#define	MAX_FILENAME_LENGTH	200    // max allowed name for files
#define MAX_FILES_BROWSE        1000  // max files to browse for exercises
#define MAX_AUXILIARY_FILES     10    // max auxiliary files
#define MAX_EXERCISES_EXAM      20    // max exercices in an exam
#define MAX_BLOCKS_EXERCISE     20    // max blocks in an exercise
#define MAX_LINESTATEMENTS_EXERCISE 100 // max lines in exercise statements
#define MAX_LINESTATEMENTS_QUESTION 50 // max lines in question statements
#define MAX_LINESTATEMENTS_ANSWER 50 // max lines in answer statements
#define MAX_QUESTIONS_BLOCK     20    // max questions in a block
#define MAX_OPTIONS_QUESTION    15    // max choices in multiplechoice
#define MAX_INIFIELDS_FORM      15    // max initial fields for the web form


// The following macro prepends the basedir path to a filename
static	char	__REL__[1000];
#define	REL(s)	(sprintf(__REL__,"%s/%s", basedir, s),__REL__)


// definitions for bison
extern  int     exerlineno;
extern	FILE	*exerin;

extern  int     desclineno;
extern	FILE	*descin;

// file handels 
EXTERN FILE    *llog,\
  *out,              \
  *corr,             \
  *ans,
  *outw,
  *answ; // (llog=logfile, out=latex output, corr: anwer key, 
         // ans= latex answers, outw=web output, answ=web answers

// to control if we are creating LaTeX or web output
EXTERN int modenormal;
EXTERN int modeweb;

// a string for multiple dirs
EXTERN int checkdirw
#ifdef MAIN_PROGRAM
=0 
#endif
;

// some strings to hold parsed statements and other tasks
EXTERN	char title[MAX_STR_LENGTH];
EXTERN	char temp[MAX_STR_LENGTH];
EXTERN  char temp2[MAX_STR_LENGTH];

// strings to hold a lot of text for general use
EXTERN	char	line[MAX_OPTIONS_QUESTION][10*MAX_STR_LENGTH];

// structure to hold questions of a block in an exercise

// Base directory
EXTERN	char basedir[MAX_STR_LENGTH]
#ifdef MAIN_PROGRAM
="." // default
#endif
;

// Documentclass options
EXTERN	char documentclassoptions[MAX_STR_LENGTH]
#ifdef MAIN_PROGRAM
="12pt,a4paper" //default
#endif
;

// Master latex file (a file with extension ".sty" has to be created)
EXTERN	char mastertex[MAX_FILENAME_LENGTH]
#ifdef MAIN_PROGRAM
="master" // default
#endif
;


// Packages included in latex file
EXTERN	char usepackage[MAX_STR_LENGTH]
#ifdef MAIN_PROGRAM
=" " // default
#endif
;

// Number of master files included
// Fields for formweb
EXTERN  int masterfilenum
#ifdef MAIN_PROGRAM
=0 // initial value
#endif
;

// Exercise prefix for database of exercises
EXTERN char exerprefix[MAX_STR_LENGTH]
#ifdef MAIN_PROGRAM
="exer" // default
#endif
;

// Title for the html page
EXTERN	char pagetitle[MAX_STR_LENGTH]
#ifdef MAIN_PROGRAM
="Exam" // default
#endif
;

// Action for formweb
EXTERN	char actionfield[MAX_STR_LENGTH];

EXTERN  int solution
#ifdef MAIN_PROGRAM
=0 // flag to see if solutions are to be included in the form
#endif
;

// Fields for formweb
EXTERN  int inifieldnum
#ifdef MAIN_PROGRAM
=0 // initial value
#endif
;

EXTERN char namefield[MAX_INIFIELDS_FORM][MAX_FILENAME_LENGTH];
EXTERN char labelfield[MAX_INIFIELDS_FORM][MAX_FILENAME_LENGTH];
EXTERN char typefield[MAX_INIFIELDS_FORM][MAX_FILENAME_LENGTH];
EXTERN char sizefield[MAX_INIFIELDS_FORM][MAX_FILENAME_LENGTH];
EXTERN char textfield[MAX_INIFIELDS_FORM][MAX_STR_LENGTH];
EXTERN char textrequired[MAX_INIFIELDS_FORM][MAX_STR_LENGTH];
EXTERN char fieldname[MAX_INIFIELDS_FORM][MAX_STR_LENGTH];
EXTERN char printfield[MAX_INIFIELDS_FORM][MAX_STR_LENGTH];
EXTERN char printtextreq[MAX_INIFIELDS_FORM][MAX_STR_LENGTH];
EXTERN int  fieldrequired[MAX_INIFIELDS_FORM];
EXTERN int  checkrequired
#ifdef MAIN_PROGRAM
=0 // initial value
#endif
;
EXTERN char uncheck[MAX_STR_LENGTH];
EXTERN char textsubmit[MAX_STR_LENGTH];

// File prefix to output exams (the program will append ".tex")
EXTERN	char examtex[MAX_FILENAME_LENGTH]
#ifdef MAIN_PROGRAM
="exam" // default
#endif
;

// File prefix to output exams (the program will append ".tex")
EXTERN	char answertex[MAX_FILENAME_LENGTH]
#ifdef MAIN_PROGRAM
="answer" // default
#endif
;

// random seed
EXTERN	int	seed
#ifdef	MAIN_PROGRAM
=12345 // default
#endif
;

// various counters
EXTERN  int  i
#ifdef	MAIN_PROGRAM
=0 
#endif
;
EXTERN  int  j
#ifdef	MAIN_PROGRAM
=0 
#endif
;
EXTERN  int  k
#ifdef	MAIN_PROGRAM
=0 
#endif
;
EXTERN int chran // to be used to cycle over exerp to get random files
#ifdef	MAIN_PROGRAM
=0
#endif
;
EXTERN int numrequired // to be used to count the number of required fields
#ifdef	MAIN_PROGRAM
=0
#endif
;

// various options set in the exercise files
EXTERN int type_block; // will be 1="M", 2="S" or 3="T" 
EXTERN int truefalse; // 0=true, 1=false

EXTERN char label_tf[ ][MAX_STR_LABEL_TF]
#ifdef MAIN_PROGRAM
={"Choose","True","False"}
#endif
; // to hold labels of truefalse

EXTERN	int	numexam;  // number of exams to be created
EXTERN  int     exam
#ifdef	MAIN_PROGRAM
=0 // exam number being created (serial)
#endif
;
EXTERN  int  exer
#ifdef	MAIN_PROGRAM
=0 // count for total exercises
#endif
;
EXTERN  int  realexer
#ifdef	MAIN_PROGRAM
=0 // count for real total exercises (using EXERCISE ENDEXERCISE);
#endif
;
EXTERN  int blocknum
#ifdef	MAIN_PROGRAM
=0 // block number within exercise
#endif
;
EXTERN  int     numexer;  // number of exercises in exam
EXTERN  int     singleexer
#ifdef	MAIN_PROGRAM
=0 // to use with EXERCISE and ENDEXERCISE
#endif
;
EXTERN  int     numtype;  // number of types
EXTERN  int     numdir[MAX_EXERCISES_EXAM]; // number of dirs for a given type
EXTERN  int     iniserial
#ifdef	MAIN_PROGRAM
=1 // initial serial number of papers 
#endif
;

EXTERN  int     countques; // number of questions
EXTERN  int     utilcount // a general int to pass info between subroutines
#ifdef	MAIN_PROGRAM
=0 // block number within exercise
#endif
;

// array to hold names of dirs with questions (max dirs to browse: 20)
EXTERN char type[MAX_EXERCISES_EXAM][20][MAX_FILENAME_LENGTH];
EXTERN int numqtype[MAX_EXERCISES_EXAM]; // number of exercises for a type
EXTERN int exeqtype[MAX_EXERCISES_EXAM]; // single exercise yes or no
EXTERN int exeqput[MAX_EXERCISES_EXAM]; // check to put exercise group header
EXTERN int elemp[MAX_EXERCISES_EXAM][20][1000]; // to hold random numbers

// various arrays to hold the full exam before outputting it
EXTERN char examexertitle[MAX_EXERCISES_EXAM][MAX_STR_LENGTH]; // title
EXTERN int examexertitlech[MAX_EXERCISES_EXAM]; // check if there is a title

EXTERN int examexersingle[MAX_EXERCISES_EXAM]; // single exercise

EXTERN int examexernumst[MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]; // number of statement lines
EXTERN int examexernumst2[MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]; // number of statement lines
                            // this is one is used to keep the info
EXTERN char examexerstlin\
     [MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]\
     [MAX_LINESTATEMENTS_EXERCISE]\
     [MAX_STR_LENGTH]; // the actual statement lines of each exercise
EXTERN	int  numexerstlin
#ifdef	MAIN_PROGRAM
=0 // number of statement lines in exercise, counter used in desc.y
#endif
;

EXTERN int examexerblocks[MAX_EXERCISES_EXAM]; // number of blocks in exercise

EXTERN int examexerbltype\
       [MAX_EXERCISES_EXAM]\
       [MAX_BLOCKS_EXERCISE]; // type of block, 1=multiplechoice,
                              // 2=shortanswer and 3=truefalse

EXTERN int exexblquto\
       [MAX_EXERCISES_EXAM]\
       [MAX_BLOCKS_EXERCISE]; // number of questions in block

EXTERN int examexernumqu[MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]\
     [MAX_QUESTIONS_BLOCK]; // number of question lines
EXTERN char examexerqulin\
     [MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]\
     [MAX_QUESTIONS_BLOCK]\
     [MAX_LINESTATEMENTS_QUESTION]\
     [MAX_STR_LENGTH]; // the actual question lines of each exercise

EXTERN int examexernuman[MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]\
     [MAX_QUESTIONS_BLOCK]; // number of answer lines
EXTERN char examexeranlin\
     [MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]\
     [MAX_QUESTIONS_BLOCK]\
     [MAX_LINESTATEMENTS_ANSWER]\
     [MAX_STR_LENGTH]; // the actual answer lines of each exercise
EXTERN	int  numexerqulin
#ifdef	MAIN_PROGRAM
=0 // number of question lines in exercise, counter used in exer.y
#endif
;EXTERN	int  numexeranlin
#ifdef	MAIN_PROGRAM
=0 // number of answer lines in exercise, counter used in exer.y
#endif
;
EXTERN int exerblockqueanch[MAX_EXERCISES_EXAM]\
     [MAX_BLOCKS_EXERCISE]\
     [MAX_QUESTIONS_BLOCK]; // to check if there is an answer at all

EXTERN int examexerblockquetotopt\
       [MAX_EXERCISES_EXAM]\
       [MAX_BLOCKS_EXERCISE]\
       [MAX_QUESTIONS_BLOCK]; // number of options in multiplechoiec
EXTERN int examexerblockquecorr\
       [MAX_EXERCISES_EXAM]\
       [MAX_BLOCKS_EXERCISE]\
       [MAX_QUESTIONS_BLOCK]; // correct answer in question 
EXTERN char examexerblockqueopt\
       [MAX_EXERCISES_EXAM]\
       [MAX_BLOCKS_EXERCISE]\
       [MAX_QUESTIONS_BLOCK]\
       [MAX_OPTIONS_QUESTION]\
       [MAX_STR_LENGTH]; // the actual options in multiplechoice question								   

